src/controllers/krate/search.rs

```diff
@@ -92,9 +93,25 @@ pub async fn search(app: AppState, req: Parts) -> AppResult<Json<Value>> {
                 query = query.order(Crate::with_name(q_string).desc());
 
                 if sort == "relevance" {
-                    let q = sql::<TsQuery>("plainto_tsquery('english', ")
+                    // If the query string is not a stop word, search using `plainto_tsquery(...)`.
+                    // Else if the it is a valid keyword, search by casting it to `tsquery` with weight B(keyword).
+                    // Otherwise, search using `null::tsquery`.
```

Comment:
what happens if the query string consists of multiple terms and only a subset of them are stopwords?

Reply:
Let's consider two terms separated by either a dash or a space. 
With the following sql:
```PgSql
WITH data AS (    
  SELECT *
  FROM                                                
    (
      VALUES
        ('any-one'),
        ('any one'),
        ('one-any'),
        ('one any')           
    ) q(s)                 
)
SELECT
  s,                              
  length(                                      
    to_tsvector('english', s)        
  ),                                 
  plainto_tsquery('english', s),
  CASE WHEN (
    (
      length(
        to_tsvector('english', s)
      ) != 0
    )
  ) THEN ('plainto_tsquery')
  WHEN ('t') THEN ('cast')
  ELSE 'null' END
FROM  data;
```

```
    s    | length |  plainto_tsquery  |      case       
---------+--------+-------------------+-----------------
 any-one |      2 | 'any-on' & 'one'  | plainto_tsquery
 any one |      1 | 'one'             | plainto_tsquery
 one-any |      2 | 'one-ani' & 'one' | plainto_tsquery
 one any |      1 | 'one'             | plainto_tsquery
(4 rows)
```

we can see that all searches should be performed using the old method with`plainto_tsquery` because the `length` of `to_tsvector` is not `0`. This indicates that the results are either a single non-stopword stem (if separated by a space) or multiple non-stopword stems (if separated by a dash).

Reply:
hmm, but doesn't that mean then that the PR only solves the very specific case of searching for a single stopword?

Reply:
Unfortunately, yes, I was unaware that `keyword:` was available in the frontend, and the old issue is still open. 😓

---

https://github.com/rust-lang/crates.io/pull/9117#issuecomment-2242856370
> Oh, indeed! Perhaps we should add a help button or document somewhere to inform users about some of these search tricks?

I lean towards just using `keyword:`.  Feel free to close it as it's not planned.

Reply:
It's worth noting that if we don't limit the scope to keyword (`:B`) some stopwords like `a`, `an` or others that might frequently appear in the `readme` of multiple crates could become noise.

Reply:
> I lean towards just using `keyword:`. Feel free to close it as it's not planned.

okay, sorry for the wasted effort :-/

Reply:
No worries 🙂‍↔️